lgc/patch/Continufy.cpp

```diff
@@ -0,0 +1,205 @@
+  return newFn;
+}
+
+/// Return the levels of the caller ray-tracing stages, which will be executed after the function return.
+static unsigned getReturnedLevels(int stage) {
+  // Traversal will return to RGS.
+  if (stage == -1)
+    return 1;
+
+  RtStage rtStage = static_cast<RtStage>(stage);
+  switch (rtStage) {
+  case RtStage::ShaderStageRayGeneration:
+    llvm_unreachable("Raygen shader should not arrive here.");
+  case RtStage::ShaderStageClosestHit:
+  case RtStage::ShaderStageMiss:
+    // Traversal
+    return (1u << (unsigned)CpsLevel::Traversal);
+  case RtStage::ShaderStageCallable:
+    // CHS/Miss/Callable
+    return (1u << (unsigned)CpsLevel::ClosestHit_Miss_Callable);
+  case RtStage::ShaderStageAnyHit:
+    // IS | Traversal
+    return (1u << (unsigned)CpsLevel::Intersection | 1u << (unsigned)CpsLevel::Traversal);
+  case RtStage::ShaderStageIntersection:
+    // Traversal
+    return 1u << (unsigned)CpsLevel::Traversal;
+  default:
+    llvm_unreachable("Unknown raytracing shader type.");
+  }
+}
+
+/// Return level of the ray-tracing stage.
+static CpsLevel getLevelOfRtStage(int stage) {
+  // Traversal
+  if (stage == -1)
+    return CpsLevel::Traversal;
+
+  RtStage rtStage = static_cast<RtStage>(stage);
+  switch (rtStage) {
+  case RtStage::ShaderStageRayGeneration:
+    return CpsLevel::RayGen;
+  case RtStage::ShaderStageClosestHit:
+  case RtStage::ShaderStageMiss:
+  case RtStage::ShaderStageCallable:
+    return CpsLevel::ClosestHit_Miss_Callable;
+  case RtStage::ShaderStageAnyHit:
+    return CpsLevel::AnyHit_CombinedIntersection_AnyHit;
+  case RtStage::ShaderStageIntersection:
+    return CpsLevel::Intersection;
+  default:
+    llvm_unreachable("Unknown raytracing shader type.");
+  }
+}
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+// @returns : The preserved analyses (The analyses that are still valid after this pass)
+PreservedAnalyses Continufy::run(Module &module, ModuleAnalysisManager &analysisManager) {
+  LLVM_DEBUG(dbgs() << "Run the Continufy pass \n");
+  m_pipelineState = analysisManager.getResult<PipelineStateWrapper>(module).getPipelineState();
+  LLVMContext &context = module.getContext();
+
+  llvm_dialects::Builder builder(context);
+  SmallVector<Instruction *> tobeErased;
+
+  for (auto &fn : make_early_inc_range(module.functions())) {
+    MDNode *continufyStage = fn.getMetadata("continufy.stage");
+    Function *fnPtr = &fn;
+    std::optional<int32_t> currentRtStage;
+    if (continufyStage) {
+      fnPtr = insertCpsArguments(fn);
+      currentRtStage = mdconst::extract<ConstantInt>(continufyStage->getOperand(0))->getSExtValue();
+      CpsLevel level = getLevelOfRtStage(currentRtStage.value());
```

Comment:
Pass the `MDNode` directly into the `getLevelOfRtStage` function?

Reply:
The `getLevelOfRtStage()` will be later moved to some other common place like `shared/continuation`, the `MDNode` handling is quite specific to this pass. I think passing integer is more proper.